torch/csrc/autograd/input_metadata.cpp

```diff
@@ -66,15 +66,18 @@ at::Tensor InputMetadata::maybe_reduce(
   // careful oblivious logic as seen below
   if (is_nested_ || is_cpp_nested_tensor() || grad.is_nested() ||
       ::torch::autograd::is_cpp_nested_tensor(grad)) {
-    if (!is_same_shape(grad)) {
-      if (is_expandable_to_shape(grad)) {
-        return reduce_grad(grad);
-      } else {
-        fail();
-      }
-    } else {
+    if (maybe_expandable_to(grad)) {
+      at::_nested_assert_expandable_to_symint(
+          grad,
+          shape_as_dim_vector(),
+          incompatible_shape_error_message(i, grad).str());
```

Comment:
sum_to_size actually does this exact same check and inserts the exact same runtime assert as second time. The only reason we have this separate assert here is to be able to spit out a more informative error message.